### PR TITLE
feat: add down sample filter before detection module

### PR DESCRIPTION
**Pull request by tkimura4**
Signed-off-by: tomoya.kimura <tomoya.kimura@tier4.jp>

## Description
- add down sample filter before detection module (optional; default: off)
By inserting down sample filter, the calculation cost of voxel_based_compare_map_filter is reduced.

- use pointcloud_container for voxel_based_compare_map_filter and down_sample_filter
By using intra_process, it is possible to reduce the cost in communication of a huge point cloud.

## Tests performed
tested by rosbag. I confirmed that the detection module works both when use_down_sample_filter is true and false

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml [hunk 1]
```diff
@@ -35,13 +37,14 @@
     <arg name="image_number" value="$(var image_number)"/>
   </include> -->
 
-  <!-- Pointcloud map filter -->
+  <!-- detection preprocess module -->
```
Comment (yukkysaito) [inline launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml:40]: I prefer the name is `Pointcloud map filter`.

File: launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml [hunk 2]
```diff
@@ -35,13 +37,14 @@
...
-      <arg name="input" value="$(var input/obstacle_segmentation/pointcloud)"/>
-      <arg name="input_map" value="/map/pointcloud_map"/>
-      <arg name="output" value="pointcloud_map_filtered/pointcloud"/>
-      <arg name="distance_threshold" value="0.5"/>
+    <include file="$(find-pkg-share perception_launch)/launch/object_recognition/detection/detection_preprocess.launch.py" if="$(var use_pointcloud_map)">
+      <arg name="input_topic" value="$(var input/obstacle_segmentation/pointcloud)"/>
+      <arg name="use_intra_process" value="true"/>
+      <arg name="use_multithread" value="true"/>
+      <arg name="use_pointcloud_container" value="$(var use_pointcloud_container)"/>
+      <arg name="container_name" value="$(var pointcloud_container_name)"/>
```
Comment (yukkysaito) [inline launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml:47]: can you add about the output arg?
```
 <arg name="output" value="pointcloud_map_filtered/pointcloud"/>
```

File: launch/tier4_perception_launch/launch/object_recognition/detection/detection_preprocess.launch.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                extra_arguments=[
+                    {"use_intra_process_comms": False},
+                ],
+            )
+        )
+        return components
+
+    def create_down_sample_pipeline(self, output_topic):
+        components = []
+        down_sample_topic = "/perception/obstacle_segmentation/downsampled/pointcloud"
```
Comment (yukkysaito) [inline launch/tier4_perception_launch/launch/object_recognition/detection/detection_preprocess.launch.py:77]:
```suggestion
        down_sample_topic = "/perception/obstacle_segmentation/pointcloud_map_filtered/downsampled/pointcloud"
```

File: launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml [hunk 4]
```diff
@@ -4,14 +4,17 @@
...
-      <arg name="input" value="$(var input/obstacle_segmentation/pointcloud)"/>
-      <arg name="input_map" value="/map/pointcloud_map"/>
-      <arg name="output" value="pointcloud_map_filtered/pointcloud"/>
-      <arg name="distance_threshold" value="0.5"/>
+    <include file="$(find-pkg-share perception_launch)/launch/object_recognition/detection/detection_preprocess.launch.py" if="$(var use_pointcloud_map)">
+      <arg name="input_topic" value="$(var input/obstacle_segmentation/pointcloud)"/>
+      <arg name="use_intra_process" value="true"/>
+      <arg name="use_multithread" value="true"/>
+      <arg name="use_pointcloud_container" value="$(var use_pointcloud_container)"/>
+      <arg name="container_name" value="$(var pointcloud_container_name)"/>
```
Comment (yukkysaito) [inline launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml:17]: same comment
https://github.com/autowarefoundation/autoware.universe/pull/961/files#r881826704

File: launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml [hunk 5]
```diff
@@ -4,14 +4,17 @@
   <arg name="input/obstacle_segmentation/pointcloud" default="/perception/obstacle_segmentation/pointcloud"/>
   <arg name="lidar_detection_model" default="centerpoint" description="options: `centerpoint`, `apollo`"/>
   <arg name="use_pointcloud_map" default="true" description="use pointcloud map in detection"/>
+  <arg name="use_pointcloud_container" default="false" description="use pointcloud container for detection preprocessor"/>
+  <arg name="container_name" default=""/>
 
-  <!-- Pointcloud map filter -->
+  <!-- detection preprocess module -->
```
Comment (yukkysaito) [inline launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml:10]: same comment
https://github.com/autowarefoundation/autoware.universe/pull/961/files#r881825935

File: launch/tier4_perception_launch/launch/object_recognition/detection/detection_preprocess.launch.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
+# Copyright 2022 Tier IV, Inc. All rights reserved.
```
Comment (yukkysaito) [inline launch/tier4_perception_launch/launch/object_recognition/detection/detection_preprocess.launch.py:1]:
```suggestion
# Copyright 2022 TIER IV, Inc. All rights reserved.
```

File: launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml (hunk 1 above)
Comment (tkimura4) [reply launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml]: Fixed at [499266a](https://github.com/autowarefoundation/autoware.universe/pull/961/commits/499266a96b5bb4ae0be85a796c2846484fa37ee4)

File: launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml (hunk 5 above)
Comment (tkimura4) [reply launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml]: Fixed at [f0e6e5e](https://github.com/autowarefoundation/autoware.universe/pull/961/commits/f0e6e5e0f396b4c81b6c6d2e6ca85550944d4fb3)

File: launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml (hunk 4 above)
Comment (tkimura4) [reply launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml:18]: Fixed at [f0e6e5e](https://github.com/autowarefoundation/autoware.universe/pull/961/commits/f0e6e5e0f396b4c81b6c6d2e6ca85550944d4fb3)

File: launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml (hunk 2 above)
Comment (tkimura4) [reply launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml:48]: Fixed at [f0e6e5e](https://github.com/autowarefoundation/autoware.universe/pull/961/commits/f0e6e5e0f396b4c81b6c6d2e6ca85550944d4fb3)

**Review by yukkysaito — APPROVED** [review]
LGTM

**Pull request merged**